### PR TITLE
feat(tls): TLS client cert auth support

### PR DESCRIPTION
**Pull request by boxmein**
This PR adds support for setting the TLS client certificate and client private key. It is manually tested and working with TLS EC P-384 keys (for my company's MQTT broker), although any keys supported by rustls (e.g. RSA, EC, ...) should work fine.

The client certificate has to be in a PEM file, and decode to the form accepted by https://docs.rs/rustls/latest/rustls/struct.Certificate.html (DER-Encoded x509)

The client private key has to be in a PEM file, and the form accepted by https://docs.rs/rustls/latest/rustls/struct.PrivateKey.html (DER-Encoded ASN.1 in PKCS#8 form)

The CLI flags accept paths (relative or absolute). Equivalent env vars are accepted as well`.

```
      --client-cert <CLIENT_CERT_PATH>
          Client cert path

          [env: MQTTUI_CLIENT_CERTIFICATE=]

      --client-key <CLIENT_PRIVATE_KEY>
          Client private key path

          [env: MQTTUI_CLIENT_PRIVATE_KEY=]
```

Todos: 

- [ ] Better docstrings
- [ ] Code review

### CODE REVIEW

File: Cargo.toml
Comment (EdJoPaTo) [inline Cargo.toml:41]: Personally I started using the major component of semver in the cargo.toml and let the lock file handle the specifics. dependency updates only require changing the lock file then

```suggestion
rustls-pemfile = { version = "1", optional = true }
```

File: src/cli.rs
Comment (EdJoPaTo) [inline src/cli.rs:172]: I think I would either rename the variable or the long attribute. Either its already clear in the short form then its that way both on CLI and in code or it needs to be more specific then it should be more specific both in code and CLI.

Also the current long form is not as specific as the env currently, this should also be aligned.

Maybe go with `client_key` and keep the doc comment saying its the private key?

File: src/cli.rs
Comment (EdJoPaTo) [inline src/cli.rs:163]: Same here, I would try to keep the long and variable name the same.

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]
I havnt really looked into the code yet but I already have some comments.

Somehow I feel like there should be a library in between mqttui and rumqttc to handle CLI parameters. Then all tools using CLI and MQTT would benefit from these kind of improvements. But thats a different topic for another time I guess.

**Comment by boxmein** [comment]
@EdJoPaTo Thanks so much for the review! 

I've taken a look at the recommendations and implemented related fixes in the command line. Also hopefully the docstrings are a bit better now.

As for the update regarding the library between CLI and rumqttc - I agree, but also not really sure how to continue on that path.

What do you think are the next steps with this code change?

**Comment by boxmein** [comment]
I also took a look at the CI checks and fixed up fmt and clippy :)

**Comment by EdJoPaTo** [comment]
Looks like you disabled me pushing to your Pull Request. I pushed my merge conflict fix and lint fixed to this branch: https://github.com/EdJoPaTo/mqttui/tree/boxmein-pr-99

File: src/cli.rs
Comment (EdJoPaTo) [inline src/cli.rs:160]: `long` assumes the name from the variable, which is less error prone afterwards to forget to rename something afterwards. Sorry for not writing that after my last change request here.

```suggestion
        long,
```

File: src/cli.rs
Comment (EdJoPaTo) [inline src/cli.rs:171]:
```suggestion
        long,
```

File: src/main.rs
Comment (EdJoPaTo) [inline src/main.rs:6]: I moved the `std` import up in bc60f94f250574db6e5c3f4aa34dde2540ad944b which creates a merge confict. Tried to fix it for you but don't seem to have push access to the PR. Just as a heads up that that's the reason for the merge conflict so you have an easier time to fix it :)

File: src/mqtt/encryption.rs
Comment (EdJoPaTo) [inline src/mqtt/encryption.rs:79]: Both of these lines are using `println!()` a lot. I am not sure if I like this default.
When using the interactive view its not seen anyway and when using the other commands it might create additional output besides what is actually wanted.

Not sure how to deal with it. Maybe using a `--verbose` flag? Maybe with an additional subcommand like `test-connection` or something like that which is especially verbose while connecting. Any thoughts?

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]

**Comment by EdJoPaTo** [comment]
Sorry for being a bit pedantic, but I prefer to do things a bit more right than rushed. Then I am happy to merge your improvements!

**Comment by boxmein** [comment]
> Sorry for being a bit pedantic but I prefer to do things a bit more right than rushed. Then I am happy to merge your improvements!

That's perfectly fine. It's your project, so it's your say if the code is right!

**Comment by boxmein** [comment]
> Looks like you disabled me pushing to your Pull Request. I pushed my merge conflict fix and lint fixed to this branch: https://github.com/EdJoPaTo/mqttui/tree/boxmein-pr-99
> 
> When opening a PR there should be a checkbox to allow maintainers to push to the Pull Request. But it might also not work because its from the main branch and not a dedicated patch branch? Not entirely sure how it works.

Not really sure, I think it's because I was working directly on `main` in my own repo.

File: src/mqtt/encryption.rs
Comment (boxmein) [reply src/mqtt/encryption.rs:79]: Sorry, debugging lines should not really be there before merge

File: src/main.rs
Comment (boxmein) [reply src/main.rs:6]: Likely that's because I was working on `main`

**Comment by boxmein** [comment]
I'm fixing up my branching to make sure you can commit as well

**Comment by boxmein** [comment]
#102 reopens this PR with a changed head branch

File: src/main.rs
Comment (boxmein) [reply src/main.rs]: Addressed in https://github.com/EdJoPaTo/mqttui/pull/102/commits/3042f54b11aa4c06388797a8191d41d7f9733e75

**Comment by EdJoPaTo** [comment]
The commits will be squashed anyway so as long as there are no conflicts, merges and so on aren't a problem. But thanks for trying to make things simpler! :)

**Pull request closed without merging**
